<a href="https://colab.research.google.com/github/mavillot/FUNSD-Entity-Linking/blob/main/LayoutLM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Modelo de lenguaje

## Data preparation

### Dataset:

In [ ]:
%%capture
!wget https://guillaumejaume.github.io/FUNSD/dataset.zip -O dataset.zip
!unzip dataset.zip

### Extracting candidates 

In [ ]:
import json
from pathlib import Path
import cv2
import numpy as np

In [ ]:
def candidates(path):
  anot=json.loads(open(path).read())
  img=Path(path).stem+'.png'
  path_image=Path(path).parent.parent/'images'/img
  h,w,_=cv2.imread(str(path_image)).shape
  question, answer=[],[]
  text={}
  for block in anot['form']:
    if block['label']=='question':
      question.append([block['id'],block['box']])
      text[block['id']]=block['text']
    if block['label']=='answer':
      answer.append([block['id'],block['box']])
      text[block['id']]=block['text']
  # We have built 2 list with answers and questions' info 
  #For each answer we look for its question
  dic={}
  for a in answer:
    bbox=a[1]
    candidates=[]
    candidates_more=[]
    x_a=int(bbox[0])
    x1_a=int(bbox[2])
    y_a=int(bbox[1])
    y1_a=int(bbox[3])
    pto_a=[(x_a+x1_a)/2,(y_a+y1_a)/2]
    for q in question:
      bbox=q[1]
      x_q=int(bbox[0])
      x1_q=int(bbox[2])
      y_q=int(bbox[1])
      y1_q=int(bbox[3])
      pto_q=[(x_q+x1_q)/2,(y_q+y1_q)/2]
      if x_q<x1_a+0.05*w and y1_q>y_a-0.1*h and y_q<y1_a+0.01*h:
        dist=np.sqrt((x_a-x1_q)**2+(pto_a[1]-pto_q[1])**2)
        candidates.append([q[0],dist])
      if x_q<x1_a +0.05*w and y1_q>y_a-0.6*h and y_q<y1_a+0.03*h:
        dist=np.sqrt((x_a-x1_q)**2+(pto_a[1]-pto_q[1])**2)
        candidates_more.append([q[0],dist])
    if candidates!=[]:
      dic[a[0]]=candidates
    else:
      dic[a[0]]=candidates_more
  return dic,text

In [ ]:
candidates('dataset/training_data/annotations/0000971160.json')

({4: [[1, 286.05287972680856],
   [3, 41.593268686170845],
   [10, 309.0457085934053],
   [13, 20.006249023742555],
   [15, 293.0959740426334]],
  11: [[3, 303.0264014900352], [10, 11.10180165558726]],
  12: [[3, 25.243811122728676], [10, 339.1194479825656]],
  14: [[1, 11.10180165558726],
   [10, 50.835519078691426],
   [15, 18.33712082089225]],
  16: [[1, 82.76472678623425],
   [2, 39.96248240537617],
   [3, 349.7802881810237],
   [10, 116.84177335182824],
   [13, 311.4498354470588],
   [15, 78.18567643756751]],
  18: [[17, 26.419689627245813]],
  21: [[5, 15.402921800749363],
   [6, 254.32656172724074],
   [7, 333.7813655673426],
   [20, 21.587033144922902]],
  22: [[5, 118.71394189394942], [20, 112.58885380000989]]},
 {1: ':',
  2: 'Suggestion:',
  3: 'Date:',
  4: 'Licensee',
  5: '',
  6: 'Yes',
  7: 'No',
  10: 'Name / Phone Ext. :',
  11: 'M. Hamann P. Harper, P. Martinez',
  12: '9/ 3/ 92',
  13: 'R&D Group:',
  14: 'J. S. Wigand',
  15: 'Supervisor / Manager',
  16: 'Disconti

We rank the candidates according to the distance

In [ ]:
from operator import itemgetter

In [ ]:
def rank(path):
  cand,text=candidates(path)
  for c in cand:
    aux=cand[c]
    cand[c]=[l[0] for l in sorted(aux, key=itemgetter(1))]
  return cand,text

In [ ]:
rank('dataset/training_data/annotations/0000971160.json')

({4: [13, 3, 1, 15, 10],
  11: [10, 3],
  12: [3, 10],
  14: [1, 15, 10],
  16: [2, 15, 1, 10, 13, 3],
  18: [17],
  21: [5, 20, 6, 7],
  22: [20, 5]},
 {1: ':',
  2: 'Suggestion:',
  3: 'Date:',
  4: 'Licensee',
  5: '',
  6: 'Yes',
  7: 'No',
  10: 'Name / Phone Ext. :',
  11: 'M. Hamann P. Harper, P. Martinez',
  12: '9/ 3/ 92',
  13: 'R&D Group:',
  14: 'J. S. Wigand',
  15: 'Supervisor / Manager',
  16: 'Discontinue coal retention analyses on licensee submitted product samples (Note : Coal Retention testing is not performed by most licensees. Other B&W physical measurements as ends stability and inspection for soft spots in ciparettes are thought to be sufficient measures to assure cigarette physical integrity. The proposed action will increase laboratory productivity . )',
  17: 'Suggested Solutions (s) :',
  18: 'Delete coal retention from the list of standard analyses performed on licensee submitted product samples. Special requests for coal retention testing could still be sub

We construct now a function that given a path it returns for each answer the candidates, the labels and the text of the questions and answers

In [ ]:
def candidatesANDlabels(path):
  anot=json.loads(open(path).read())
  cand,text=rank(path)
  dic_label={}
  for block in anot['form']:
    if block['label']=='answer':
      id=block['id']
      id_question=[]
      for link in block['linking']:
        if link[0]==id:
          id_question.append(link[1])
        else: id_question.append(link[0])
      lista=cand[id]
      labels=[lista[i] in id_question for i in range(0,len(lista))]
      dic_label[id]=labels
  return cand,dic_label,text     

In [ ]:
candidatesANDlabels('dataset/training_data/annotations/0000971160.json')

({4: [13, 3, 1, 15, 10],
  11: [10, 3],
  12: [3, 10],
  14: [1, 15, 10],
  16: [2, 15, 1, 10, 13, 3],
  18: [17],
  21: [5, 20, 6, 7],
  22: [20, 5]},
 {4: [True, False, False, False, False],
  11: [True, False],
  12: [True, False],
  14: [False, True, False],
  16: [True, False, False, False, False, False],
  18: [True],
  21: [False, True, False, False],
  22: [True, False]},
 {1: ':',
  2: 'Suggestion:',
  3: 'Date:',
  4: 'Licensee',
  5: '',
  6: 'Yes',
  7: 'No',
  10: 'Name / Phone Ext. :',
  11: 'M. Hamann P. Harper, P. Martinez',
  12: '9/ 3/ 92',
  13: 'R&D Group:',
  14: 'J. S. Wigand',
  15: 'Supervisor / Manager',
  16: 'Discontinue coal retention analyses on licensee submitted product samples (Note : Coal Retention testing is not performed by most licensees. Other B&W physical measurements as ends stability and inspection for soft spots in ciparettes are thought to be sufficient measures to assure cigarette physical integrity. The proposed action will increase laborator

### Dataset Class

We create this class in order to obtain the dataset we will use for training the BERT model. 
For each form we extract all the answers and its candidates for question. For each candidate we concatenate its text with the answer text. If the link between those 2 entities exists the label will be: 1 and if it doesn't exist it will be 0.

In [ ]:
import os
class Dataset():
    def __init__(self, path_annotation):
        self.path_annotation= path_annotation

    def __iter__(self):
      with os.scandir(self.path_annotation) as files:
        for file in files:
          yield file.name
          
    def __len__(self):
      i=0
      for file in self:
        i+=1
      return i

    def textList(self, o):
      path=self.path_annotation+'/'+o
      cand,label,text=candidatesANDlabels(path)
      question_answer,et=[],[]
      dic={True:1,False:0}
      for c in cand:
        question_answer=question_answer+[text[x]+' '+text[c] for x in cand[c]]
        et=et+[dic[z] for z in label[c]]
      return question_answer,et
    
    def preparation(self):
      text=[]
      label=[]
      for file in self:
        txt,lbl=self.textList(file)
        text=text+txt
        label=label+lbl
      return (text,label)

In [ ]:
dataset_train=Dataset('dataset/training_data/annotations')
dataset_test=Dataset('dataset/testing_data/annotations')

In [ ]:
text_train, labels_train=dataset_train.preparation()
text_test, labels_test=dataset_test.preparation()

In [ ]:
from sklearn.model_selection import train_test_split
train_texts, val_texts, train_labels, val_labels = train_test_split(text_train, labels_train, test_size=.2)

**We check if all the answers have at least one question in their candidates:**

In [ ]:
path='/content/dataset/training_data/annotations/'
i=0
for f in dataset_train:
  f=path+f
  cand,_,_=candidatesANDlabels(f)
  for c in cand:
    if cand[c]==[]:
      i+=1
print(i)

0


There's no answer without candidates

**We check if all the correct questions are in the candidate list of the answer:**

In [ ]:
path='/content/dataset/training_data/annotations/'
i,j=0,0
for f in dataset_train:
  f=path+f
  cand,lbl,text=candidatesANDlabels(f)
  for k in lbl:
    j+=1
    if not(True in lbl[k]):
      i+=1
print(i,j)

47 2802


In [ ]:
47*100/2802

1.6773733047822983

Ther're 47 of the 2802 answers (1.68%) who haven't got their correct question in the candidates

### FUNSDataset Class

This class helps us to pass the dataset in a format that the model understands.

In [ ]:
import torch
class FUNDSDataset(torch.utils.data.Dataset):
    def __init__(self, encodings, labels):
        self.encodings = encodings
        self.labels = torch.tensor(labels)

    def __getitem__(self, idx):
        item = {key: torch.tensor(val[idx]) for key, val in self.encodings.items()}
        item['labels'] = torch.tensor(self.labels[idx])
        return item

    def __len__(self):
        return len(self.labels)

## Modelo

In [ ]:
%%capture
pip install transformers

#### Metrics

These are the metrics we will be using for training.

In [ ]:
from sklearn.metrics import accuracy_score, precision_recall_fscore_support

def compute_metrics(pred):
    labels = pred.label_ids
    preds = pred.predictions.argmax(-1)
    precision, recall, f1, _ = precision_recall_fscore_support(labels, preds, average='binary')
    acc = accuracy_score(labels, preds)
    return {
        'accuracy': acc,
        'f1': f1,
        'precision': precision,
        'recall': recall
    }

##### mAP

In [ ]:
import numpy as np
from sklearn.metrics import average_precision_score

def mAP_x(scores, labels):
  m=average_precision_score(labels, scores)
  if np.isnan(m):
    return 0
  else:
    return m

In [ ]:
def mAP(dataset,coef_FINAL,labels):
  i=0
  map=[]
  for f in dataset:
    cand,txt=rank(dataset.path_annotation+ '/' +f)
    n=len(cand)
    if n>0:
      for c in cand:
        j=i+len(cand[c])
        map.append(mAP_x(coef_FINAL[i:j],labels[i:j]))
        i=j
  return sum(map)/len(map)

##### mRank

In [ ]:
import numpy as np
def mRank_x(predictions, labels,coef):
  if predictions==labels:
    return 0
  else:
    if 1 in labels:
      k=labels.index(1)
      return sum([c>coef[k] for c in coef])
    else:
      return len(predictions)

In [ ]:
def mRank(dataset,prediction,labels,coef):
  mrank=[]
  i=0
  for f in dataset:
    cand,txt=rank(dataset.path_annotation+ '/' +f)
    n=len(cand)
    if n>0: #comprobamos que el documento tenga algun par clave-valor
      for c in cand:
        j=i+len(cand[c])
        if j!=i:
          mrank.append(mRank_x(prediction[i:j],labels[i:j],coef[i:j]))
          i=j
  return sum(mrank)/len(mrank)

#### Tokenizer and model

In [ ]:
from transformers import BertForSequenceClassification,AutoTokenizer,Trainer, TrainingArguments
tokenizer = AutoTokenizer.from_pretrained('microsoft/layoutlm-base-uncased')
model = BertForSequenceClassification.from_pretrained('microsoft/layoutlm-base-uncased', num_labels=2)

Downloading:   0%|          | 0.00/170 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/606 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/226k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/455k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/112 [00:00<?, ?B/s]

You are using a model of type layoutlm to instantiate a model of type bert. This is not supported for all configurations of models and can yield errors.


Downloading:   0%|          | 0.00/432M [00:00<?, ?B/s]

Some weights of the model checkpoint at microsoft/layoutlm-base-uncased were not used when initializing BertForSequenceClassification: ['layoutlm.encoder.layer.6.attention.self.value.bias', 'layoutlm.embeddings.word_embeddings.weight', 'layoutlm.encoder.layer.3.attention.output.LayerNorm.bias', 'layoutlm.encoder.layer.6.attention.self.key.weight', 'layoutlm.encoder.layer.2.attention.self.query.weight', 'layoutlm.encoder.layer.10.attention.self.key.bias', 'layoutlm.encoder.layer.10.output.dense.weight', 'layoutlm.encoder.layer.6.attention.output.dense.bias', 'layoutlm.encoder.layer.2.attention.self.key.weight', 'layoutlm.encoder.layer.8.output.LayerNorm.weight', 'layoutlm.embeddings.y_position_embeddings.weight', 'layoutlm.encoder.layer.11.attention.output.dense.weight', 'layoutlm.embeddings.token_type_embeddings.weight', 'layoutlm.encoder.layer.8.attention.self.key.weight', 'layoutlm.encoder.layer.6.attention.self.query.bias', 'layoutlm.encoder.layer.0.attention.output.dense.weight', '

In [ ]:
train_encodings = tokenizer(train_texts, truncation=True, padding=True)
val_encodings = tokenizer(val_texts, truncation=True, padding=True)
test_encodings = tokenizer(text_test, truncation=True, padding=True)

In [ ]:
train_dataset = FUNDSDataset(train_encodings, train_labels)
val_dataset = FUNDSDataset(val_encodings, val_labels)
test_dataset = FUNDSDataset(test_encodings, labels_test)

#### Training arguments:

In [ ]:
args = TrainingArguments(
    output_dir='./results',
    evaluation_strategy = "epoch",
    save_strategy='epoch',
    learning_rate=2e-5,
    per_device_train_batch_size=32,
    per_device_eval_batch_size=16,
    num_train_epochs=6,
    weight_decay=0.01,
    load_best_model_at_end=True,
    metric_for_best_model='f1',
)

In [ ]:
trainer = Trainer(
    model,
    args,
    train_dataset=train_dataset,
    eval_dataset=val_dataset,
    tokenizer=tokenizer,
    compute_metrics=compute_metrics
)

#### Training

In [ ]:
trainer.train() 

***** Running training *****
  Num examples = 10502
  Num Epochs = 6
  Instantaneous batch size per device = 32
  Total train batch size (w. parallel, distributed & accumulation) = 32
  Gradient Accumulation steps = 1
  Total optimization steps = 1974
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:9: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  if __name__ == '__main__':


***** Running Evaluation *****
  Num examples = 2626
  Batch size = 16
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
Saving model checkpoint to ./results/checkpoint-329
Configuration saved in ./results/checkpoint-329/config.json
Model weights saved in ./results/checkpoint-329/pytorch_model.bin
tokenizer config file saved in ./results/checkpoint-329/tokenizer_config.json
Special tokens file saved in ./results/checkpoint-329/special_tokens_map.json
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:9: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  if __name__ == '__main__':
***

TrainOutput(global_step=1974, training_loss=0.4957867886157746, metrics={'train_runtime': 2963.5434, 'train_samples_per_second': 21.262, 'train_steps_per_second': 0.666, 'total_flos': 4565743532555760.0, 'train_loss': 0.4957867886157746, 'epoch': 6.0})

In [ ]:
trainer.train() 

***** Running training *****
  Num examples = 10502
  Num Epochs = 6
  Instantaneous batch size per device = 32
  Total train batch size (w. parallel, distributed & accumulation) = 32
  Gradient Accumulation steps = 1
  Total optimization steps = 1974
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:9: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  if __name__ == '__main__':


***** Running Evaluation *****
  Num examples = 2626
  Batch size = 16
Saving model checkpoint to ./results/checkpoint-329
Configuration saved in ./results/checkpoint-329/config.json
Model weights saved in ./results/checkpoint-329/pytorch_model.bin
tokenizer config file saved in ./results/checkpoint-329/tokenizer_config.json
Special tokens file saved in ./results/checkpoint-329/special_tokens_map.json
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:9: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  if __name__ == '__main__':
***** Running Evaluation *****
  Num examples = 2626
  Batch size = 16
Saving model checkpoint to ./results/checkpoint-658
Configuration saved in ./results/checkpoint-658/config.json
Model weights saved in ./results/checkpoint-658/pytorch_model.bin
tokenizer config file saved in ./results/checkpoint-658/t

TrainOutput(global_step=1974, training_loss=0.44529796009005745, metrics={'train_runtime': 2951.4854, 'train_samples_per_second': 21.349, 'train_steps_per_second': 0.669, 'total_flos': 4565743532555760.0, 'train_loss': 0.44529796009005745, 'epoch': 6.0})

Saving the model

In [ ]:
trainer.save_model('layoutlm')

Saving model checkpoint to layoutlm
Configuration saved in layoutlm/config.json
Model weights saved in layoutlm/pytorch_model.bin
tokenizer config file saved in layoutlm/tokenizer_config.json
Special tokens file saved in layoutlm/special_tokens_map.json


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import shutil
shutil.make_archive('/content/drive/MyDrive/layoutlm', 'zip', 'layoutlm')

'/content/drive/MyDrive/layoutlm.zip'

## Evaluation

In [ ]:
trainer.evaluate(test_dataset)

***** Running Evaluation *****
  Num examples = 2110
  Batch size = 16
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:9: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  if __name__ == '__main__':


{'epoch': 6.0,
 'eval_accuracy': 0.5587677725118484,
 'eval_f1': 0.43949428055388323,
 'eval_loss': 0.6979271769523621,
 'eval_precision': 0.3798126951092612,
 'eval_recall': 0.5214285714285715,
 'eval_runtime': 35.4767,
 'eval_samples_per_second': 59.476,
 'eval_steps_per_second': 3.721}

## Downloading and loading the trained weights

In [ ]:
import shutil
!wget https://github.com/mavillot/FUNSD-Entity-Linking/releases/download/layoutLM/layoutlm.zip
shutil.unpack_archive('layoutlm.zip', 'layoutlm')

In [ ]:
from transformers import BertForSequenceClassification,AutoTokenizer,Trainer, TrainingArguments
tokenizer = AutoTokenizer.from_pretrained('layoutlm')
model = BertForSequenceClassification.from_pretrained('layoutlm', num_labels=2)
trainer = Trainer(model,compute_metrics=compute_metrics)

# Results

We build our test dataset

In [ ]:
dataset_test=Dataset('dataset/testing_data/annotations')
text_test, labels_test=dataset_test.preparation()
test_encodings = tokenizer(text_test, truncation=True, padding=True)
test_dataset = FUNDSDataset(test_encodings, labels_test)

**We check if all the answers have at least one question in their candidates:**

In [ ]:
path='/content/dataset/testing_data/annotations/'
i=0
for f in dataset_test:
  f=path+f
  cand,_,_=candidatesANDlabels(f)
  for c in cand:
    if cand[c]==[]:
      i+=1
print(i)

0


There's no answer without candidates

**We check if all the correct questions are in the candidate list of the answer:**

In [ ]:
path='/content/dataset/testing_data/annotations/'
i,j=0,0
for f in dataset_test:
  f=path+f
  _,lbl,_=candidatesANDlabels(f)
  for k in lbl:
    j+=1
    if not(True in lbl[k]):
      i+=1
print(i,j)

3 821


In [ ]:
3*100/821

0.3654080389768575

Ther're 3 of the 821 answers (0.37%) who haven't got their correct question in the candidates

### PREDICTIONS

In [ ]:
coef=trainer.predict(test_dataset)[0]
prediction=coef.argmax(-1)
prediction=[p for p in prediction]
coef=[tupla[1] for tupla in coef]

### Results (Before the rules are applied)

In [ ]:
precision, recall, f1, _ = precision_recall_fscore_support(labels_test, prediction, average='binary')
acc = accuracy_score(labels_test, prediction)
print('accuracy: ',acc)
print('precision: ',precision)
print('recall: ',recall)
print('f1: ',f1)

accuracy:  0.8018490754622689
precision:  0.517208413001912
recall:  0.6525934861278649
f1:  0.5770666666666666


In [ ]:
mAP(dataset_test,coef,labels_test)

/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_ranking.py:677: RuntimeWarning: invalid value encountered in true_divide
  recall = tps / tps[-1]


0.7736070162075037

In [ ]:
mRank(dataset_test,prediction,labels_test,coef)

0.6516443361753959

### Results (After the rules are applied)

RULES:

In [ ]:
path=dataset_test.path_annotation+'/'
a=0
prediction_FINAL,coef_FINAL=[],[]
for f in dataset_test:
  cand,txt=rank(path+f)
  lenght=0
  for c in cand:
    lenght+=len(cand[c])
  b=a+lenght
  pred,cf=rules(cand, prediction[a:b],coef[a:b])
  prediction_FINAL=prediction_FINAL+pred
  coef_FINAL=coef_FINAL+cf
  a=b

In [ ]:
precision, recall, f1, _ = precision_recall_fscore_support(labels_test, prediction_FINAL, average='binary')
acc = accuracy_score(labels_test, prediction_FINAL)
print('accuracy: ',acc)
print('precision: ',precision)
print('recall: ',recall)
print('f1: ',f1)

accuracy:  0.9170414792603698
precision:  0.8026796589524969
recall:  0.7949336550060314
f1:  0.7987878787878787


In [ ]:
mAP(dataset_test,coef_FINAL,labels_test)

/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_ranking.py:677: RuntimeWarning: invalid value encountered in true_divide
  recall = tps / tps[-1]


0.8724088369338063

In [ ]:
mRank(dataset_test,prediction_FINAL,labels_test,coef_FINAL)

0.4933008526187576